### ISSUE TITLE
Allow users to set alternative default link

### ISSUE THREAD

**Issue by ARawles** [r-lib/downlit#106]
I'm currently using `pkgdown` to host all the documentation for my work's internal R packages. Currently, `downlit` just defaults to using https://rdrr.io for packages/documentation without a package-defined URL. This works fine if your package is publicly available, but because all of these packages are internal and so hosted on our internal network rather than CRAN, users just get sent to a dead link.

We do however have a consistent internal structure for the documentation of these internal packages (like https://rdrr.io but just for these packages), meaning that links could easily be generated for these internal packages.

It would be really helpful to be able to change the default link. Ideally this would be for selected packages, but I think that would be much tougher to implement.

Straight away I was thinking that implementing some kind of `downlit.default.host` option and some variable substitution would provide the functionality. For example, by setting the option to "https://my-custom-host/packages/%s/docs" and changing the `href_package_ref ()` function from:

```
    if (is_base_package(package)) {
      paste0("https://rdrr.io/r/", package)
    } else {
      paste0("https://rdrr.io/pkg/", package, "/man")
    }
```

to

```
...
    if (is_base_package(package)) {
      paste0("https://rdrr.io/r/", package)
    } else {
      sprintf(getOption("downlit.default.host", "https://rdrr.io/pkg/%s/man"), package)
    }
...
```

But this would then change every link where the parent package doesn't have a custom URL to that new host. That might be okay for some but then it's also quite limiting.

Happy to work on a PR and any thoughts would be appreciated.

**Comment by hadley** [comment]
Do all the internal packages have pkgdown sites? If so, can’t you use the standard pkgdown registration (https://pkgdown.r-lib.org/articles/linking.html)?

**Comment by ARawles** [comment]
They do have their own pkgdown sites and the URL is declared in the DESCRIPTION file, but does that auto-linking not require those packages to be installed when the site is built? I.e. downlit looks for the DESCRIPTION as a system file of the package

The pkgdown sites are generated as part of a CI process on a separate agent that also builds the package for the local CRAN-like repo, so I could make sure that the packages that are linked are installed at build time. And thinking about it, you'd imagine that most packages that are being linked in the package would likely also be declared as a dependency and so installed at build time. But I can image instances where you might want to link to a non-installed package somewhere in your site (e.g. linking to a package that you took inspiration from).

**Comment by hadley** [comment]
Ah good point. I could've sworn that `package_urls()` also fell back to looking for a DESCRIPTION on CRAN, but clearly I imagined that. Independently of this issue, it seems like that would be good idea. For your case, I wonder if it would make sense to be able to register someway way of finding the DESCRIPTION, rather than the website itself. Then downlit could look for a DESCRIPTION in the installed package, then with the custom function, then fallback to CRAN.

**Comment by ARawles** [comment]
Yeah I agree that registering a way to find the DESCRIPTION is probably a better general solution that specifying a website. Perhaps even just by changing the order of precedence to:

Check installed package -> Repositories in user's "repos" option -> Default to CRAN

That could account for all CRAN and CRAN-like custom packages. Hopefully this could then account for the vast majority of cases (including mine).

I think the ability to register a custom function to be used at some point in that process (maybe before the default to CRAN) would definitely be the most powerful and would be able to then account for (possibly) every other edge case. Any thoughts on the best way to implement that?

**Comment by hadley** [comment]
Yeah, I like your approach. Do you want to have a go at implementing it?  

Another option worth exploring would be the remotes field of the installed package, e.g. `packageDescription("dplyr")$Repository`. But I'm not sure what this holds in general, so maybe see if it's useful for the packages you're installing from your internal repo?

**Comment by ARawles** [comment]
Yeah sure, I can take a shot at it.

Might take a couple of weeks due to work but I'll keep this thread updated if anything comes up along the way 👍

**Comment by ARawles** [comment]
PR ready (#108).

With the changes, downlit will look at the local install, then search the PACKAGES files of the user's `repos` option (excluding CRAN), and then finally look on CRAN if it hasn't found anything.

This does mean though that the custom repositories will need to specifically include the URL field in their PACKAGE files, which isn't included by default (but it would be a lot easier if it was!). For me this is fine because I can make that change but I'm not sure how well this will fare for others.

For the CRAN search, we use the `tools::CRAN_package_db()` function to get package meta-data that isn't included in the PACKAGES files. This does include the URL field so checking that is nice and easy.

Ideally, there would be a uniform way of checking for the URL regardless of the repository. But because of the fact that the URL field isn't included in the PACKAGES file by default, the only way to do this would be to download the package and manually extract the DESCRIPTION file, which seems like overkill.
